examples/move/big_set/sources/big_set.move

```diff
@@ -0,0 +1,215 @@
+    }
+
+    /// Returns true if `key` is present in self.
+    public fun contains<K: copy + store + drop>(
+        self: &BigSet<K>, key: &K
+    ): bool {
+        let set_key = key(key, self.scale);
+        df::exists_with_type<Key, VecSet<K>>(&self.id, set_key)
+            && vec_set::contains(df::borrow(&self.id, set_key), key)
+    }
+
+    /// Returns the number of entries in self.
+    public fun size<K: copy + store + drop>(self: &BigSet<K>): u64 {
+        self.size
+    }
+
+    /// Returns true if self is empty.
+    public fun is_empty<K: copy + store + drop>(self: &BigSet<K>): bool {
+        self.size == 0
+    }
+
+    /// TODO: Not implemented; potentially not implementable.
+    public fun keys<K: copy + store + drop>(_self: &BigSet<K>): &vector<K> {
+        abort ENotImplemented
+    }
+
+    /// Unpacks `BigSet` into a vector of all ever stored keys.
+    public fun into_keys<K: copy + store + drop>(self: BigSet<K>): vector<K> {
+        let BigSet { id, keys: set_keys, size, scale: _ } = self;
+        let result_keys = vector[];
+
+        if (size == 0) {
+            object::delete(id);
+            return result_keys
+        };
+
+        let set_keys = vec_set::into_keys(set_keys);
+        while (vector::length(&set_keys) > 0) {
+            let set_key = vector::pop_back(&mut set_keys);
+            let set = df::remove(&mut id, set_key);
+            let keys = vec_set::into_keys(set);
+            vector::append(&mut result_keys, keys);
+        };
+
+        object::delete(id);
+        result_keys
+    }
+
+    /// Copy the `BigSet` into a new `BigSet` following `copy` ability of the
+    /// simple `VecSet`.
+    public fun copy_<K: copy + store + drop>(
+        self: &BigSet<K>, ctx: &mut TxContext
+    ): BigSet<K> {
+        let new_set = empty(self.scale, ctx);
+        let set_keys = *vec_set::keys(&self.keys);
+
+        new_set.keys = self.keys;
+        new_set.size = self.size;
+
+        while (vector::length(&set_keys) > 0) {
+            let set_key = vector::pop_back(&mut set_keys);
+            let set_copy: VecSet<K> = *df::borrow(&self.id, set_key);
+            df::add(&mut new_set.id, set_key, set_copy);
+        };
+
+        new_set
+    }
+
+    /// Drop the `BigSet` following `drop` ability of the simple `VecSet`.
+    public fun drop<K: copy + store + drop>(self: BigSet<K>) {
+        let BigSet { id, keys: set_keys, size, scale: _ } = self;
+        if (size == 0) {
+            object::delete(id);
+            return ()
+        };
+
+        let set_keys = vec_set::into_keys(set_keys);
+        while (vector::length(&set_keys) > 0) {
+            let set_key = vector::pop_back(&mut set_keys);
+            let _: VecSet<K> = df::remove(&mut id, set_key);
+        };
+
+        object::delete(id)
+    }
+
+    /// Generate a Key for the given value.
+    fun key<K>(k: &K, scale: u8): Key {
+        let value = if (scale == 1) {
+            let bytes = blake2b256(&bcs::to_bytes(k));
```

Comment:
if I understand the implementation correctly it seems to me we are going through a great effort to pick one byte. And if that distributes reasonably well we will have a set of values across multiple datastore entry instead of compacted in a single "vector".
In other words it's not clear to me what the compromises are for this implementation.
Having said so, the idea is sound and there is a need to have a stable way to distribute values. So that is cool.
Wonder if a usage of `mod` and `div` operators would allow for an easier customization.
Not sure what your thoughts are

Reply:
I had a similar thought when I was reading this PR -- my mind immediately went to [Extendible Hashing](https://en.wikipedia.org/wiki/Extendible_hashing).  What we're doing here seems to be a hard-coded version of this specialised to using the first 8 or 16 bits of the hash, but extendible hashing would allow you to spill out across multiple objects when needed, using more of the hash bits for "hot" prefixes, and keeping things compact for "cool" prefixes (which also means initially, it will behave kind of like a regular `VecSet`).